src/trusted_ca_keys/trusted_ca_keys.cc

```diff
@@ -0,0 +1,161 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+#include <fstream>
+#include <string>
+#include <cstring>
+#include <vector>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include "oslogin_utils.h"
+#include "trusted_ca_keys.h"
+
+using std::cout;
+using std::cerr;
+using std::endl;
+using std::flush;
+using std::string;
+using std::stringstream;
+using std::vector;
+using std::ofstream;
+
+using oslogin_utils::HttpGet;
+using oslogin_utils::ParseJsonToSuccess;
+using oslogin_utils::ParseJsonToEmail;
+using oslogin_utils::ParseJsonToSshKeys;
+using oslogin_utils::UrlEncode;
+using oslogin_utils::ParseJsonToCAKeys;
+
+ofstream
+CreateTmpKeysFile(const string& trusted_keys_filename, const string& trusted_keys_tmp_filename) {
+
+  ofstream trusted_keys_tmp_file(trusted_keys_tmp_filename.c_str());
+
+  if (!trusted_keys_tmp_file.is_open()) {
+    stringstream err_msg;
+
+    err_msg << "Failed!\n[ERROR] Can't open '"
+            << trusted_keys_tmp_filename << "' for writing";
+    perror(err_msg.str().c_str());
+
+    exit(EXIT_FAILURE);
+  }
+
+  if (chown(trusted_keys_tmp_filename.c_str(), 0, 0) != 0) {
+    stringstream err_msg;
+
+    err_msg << "Failed!\n[ERROR] Failed to set root:root file owner for "
+            << trusted_keys_tmp_filename;
+    perror(err_msg.str().c_str());
+    remove(trusted_keys_tmp_filename.c_str());
+
+    exit(EXIT_FAILURE);
+  }
+
+  if (chmod(trusted_keys_tmp_filename.c_str(),
+            S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH) != 0) {
+    stringstream err_msg;
+
+    err_msg << "Failed!\n[ERROR] Failed to set 644 file permissions for "
+            << trusted_keys_tmp_filename;
+    perror(err_msg.str().c_str());
+    remove(trusted_keys_tmp_filename.c_str());
+
+    exit(EXIT_FAILURE);
+  }
+
+  return trusted_keys_tmp_file;
+}
+
+void
+WriteCAKeysTmpFile(ofstream& trusted_keys_tmp_file, const vector<string> ca_keys) {
+
+  for (size_t i = 0; i < ca_keys.size(); ++i) {
+    trusted_keys_tmp_file << ca_keys[i] << endl;
+  }
+}
+
+int
+main(int argc, char* argv[]) {
+
+  string trusted_keys_filename = trusted_keys_file_def;
+  string mds_url = mds_url_def;
+  char*  mds_url_env = NULL;
+
+  if (argc >= 2) {
+    if (strncmp(argv[1], "--help", 6) == 0 ||
+        strncmp(argv[1], "-h", 2) == 0) {
+      cerr << "Usage: google_trusted_ca_keys [/path/to/trustedca.pub]\n\n"
+           << "Environment:\n"
+           << " - GOOGLE_OSLOGIN_MDS_URL=\"http://srv/path/to/certificates\"\n"
+           << "Default values:\n"
+           << " - CA keys file: /etc/ssh/trustedca.pub\n"
+           << " - API endpoint: http://169.254.169.254/computeMetadata/v1/oslogin/certificates\n";
+
+      return EXIT_FAILURE;
+    }
+    trusted_keys_filename = argv[1];
+  }
+
+  if ( (mds_url_env = getenv("GOOGLE_OSLOGIN_MDS_URL")) != NULL) {
```

Comment:
do you need this in production? is it safe? is it consistent with the code base? I understand it's useful in your development environment but not sure about production.

Reply:
I didn't use it during development (I used a hardcoded value).

What security issues do you see here?
Do you suspect that curl library would fail if instead of a URL it find something else?
This tool will be invoked from a systemd unit that is writable only by root. The file with CA keys is writable only by root. If a non-root will run a tool, nothing bad will happen, because of a lack of permissions. But, if someone has a root on a VM, there are easier ways to brake a system than playing with an environment variable.

This variable is useful for the future purposes. For example, testing a production compiled code (on a custom metadata server)

I don't see a reason to remove this so far